### ISSUE TITLE
Support for Apple's MPS [?]

### ISSUE THREAD

**Issue by escoolioinglesias** [RVC-Project/Retrieval-based-Voice-Conversion-WebUI#136]
Will MPS be supported in the future and will it be possible to train models on Apple Silicon chips? 🙏

Thank you! 🙌

**Comment by Tps-F** [comment]
Now I'm doing-

**Comment by Gitterman69** [comment]
cant wait!!!!!

**Comment by ricardo410** [comment]
hello any update on this?

**Comment by Naozumi520** [comment]
> hello any update on this?

Isn't it supported?

**Comment by ricardo410** [comment]
@Naozumi520  only for infer, not for training.

**Comment by Tps-F** [comment]
Train fixed bf1170012564463e9b9f13e7174b1eb499bdcbd2

**Comment by Naozumi520** [comment]
Yep forgot to mentioned. They haven't released in the package. I do it by copying `train_nsf_sim_cache_sid_load_pretrain.py` and `train/process_ckpt.py` from repo to my folder and replace them.

**Comment by ricardo410** [comment]
@Naozumi520 can you train? it gives me an error. 
 Retrieval-based-Voice-Conversion-WebUI-main/train/utils.py", line 206, in latest_checkpoint_path
    x = f_list[-1]
IndexError: list index out of range

**Comment by Naozumi520** [comment]
That's weird. Yes I'm able to train.

<img width="678" alt="Screenshot 2023-06-08 at 1 13 39 PM" src="https://github.com/RVC-Project/Retrieval-based-Voice-Conversion-WebUI/assets/52615455/d31664d1-85fa-448e-be55-75e846ebb81e">

**Comment by Naozumi520** [comment]
Yes. I'm using the version [updated0528v2](https://github.com/RVC-Project/Retrieval-based-Voice-Conversion-WebUI/releases/tag/updated0528v2)

**Comment by ricardo410** [comment]
> Eso es raro. Sí, puedo entrenar.
> 
> <img alt="Captura de pantalla 2023-06-08 a las 1 13 39 p. m." width="678" src="https://user-images.githubusercontent.com/52615455/244271019-d31664d1-85fa-448e-be55-75e846ebb81e.png">

it is already working, but I thought that the training would be faster, because of the characteristics of my mac, but in reality it is slower than training online.

**Comment by ricardo410** [comment]
Any idea how to solve this?

> > it is already working, but I thought that the training would be faster, because of the characteristics of my mac, but in reality it is slower than training online.
> 
> Yeah so do I! I was wondering was mps running on my graphic card or cpu. Cuz I'm intel cpu.


Any idea how to solve this?

**Comment by Tps-F** [comment]
I think one of the reasons is that some functions in Pytorch do not support mps.
and I am rewriting the code and will let you know if I find anything.

**Comment by Tps-F** [comment]
Sorry, AMD graphics cards are not supported.(from #272 )

**Comment by Naozumi520** [comment]
Wait no I'm talking about on a Mac, which supported by MPS.

https://developer.apple.com/metal/pytorch/

**Comment by Naozumi520** [comment]
> did it work for you?

Yep as I said before I'm able to do train and inference. 
However, the speed is slower than I thought as you mentioned.

**Comment by ricardo410** [comment]
> https://developer.apple.com/metal/pytorch/

I was referring to this https://developer.apple.com/metal/pytorch/

**Comment by Naozumi520** [comment]
> I was referring to this https://developer.apple.com/metal/pytorch/

No and we don't have to follow this. PyTorch have builtin support for MPS already.

**Comment by ricardo410** [comment]
> > I was referring to this https://developer.apple.com/metal/pytorch/
> 
> No and we don't have to follow this. PyTorch have builtin support for MPS already.

if you have it, why is the training so slow? it took me 19 hours to train a 12 minute dataset with 300 epochs.

**Comment by Naozumi520** [comment]
> If you have it, why is the training so slow? it took me 19 hours to train a 12 minute dataset with 300 epochs.

If we have no MPS then we won't even be able to train the model. It's likely as @Tps-F mentioned, some functions in torch required cuda are not supported by MPS yet. Like I mentioned I was required to put the flag `PYTORCH_ENABLE_MPS_FALLBACK` to `1`. This make those unsupported function use cpu as a fallback when MPS is not supported.

**Comment by an-internet-user001** [comment]
Is anyone training successfully on an Apple Silicone system AND utilizing the GPUs?

**Comment by Tps-F** [comment]
> Is anyone training successfully on an Apple Silicone system AND utilizing the GPUs? 

What device configuration do you have?

**Comment by an-internet-user001** [comment]
I have an M2 Max Mac Studio.

**Comment by an-internet-user001** [comment]
Apologies if I am behind on this issue, but it seems as though PyTorch supports MPS now, or is this not the case?  Would you need to essentially "swap" this out for the currently implemented CUDA?   (sorry if I sound a little ignorant on the topic as I am purely a hobbyist.

MPS BACKEND

mps device enables high-performance training on GPU for MacOS devices with Metal programming framework. It introduces a new device to map Machine Learning computational graphs and primitives on highly efficient Metal Performance Shaders Graph framework and tuned kernels provided by Metal Performance Shaders framework respectively.

The new MPS backend extends the PyTorch ecosystem and provides existing scripts capabilities to setup and run operations on GPU.

To get started, simply move your Tensor and Module to the mps device:

# Check that MPS is available
if not torch.backends.mps.is_available():
    if not torch.backends.mps.is_built():
        print("MPS not available because the current PyTorch install was not "
              "built with MPS enabled.")
    else:
        print("MPS not available because the current MacOS version is not 12.3+ "
              "and/or you do not have an MPS-enabled device on this machine.")

else:
    mps_device = torch.device("mps")

    # Create a Tensor directly on the mps device
    x = torch.ones(5, device=mps_device)
    # Or
    x = torch.ones(5, device="mps")

    # Any operation happens on the GPU
    y = x * 2

    # Move your model to mps just like any other device
    model = YourFavoriteNet()
    model.to(mps_device)

    # Now every call runs on the GPU
    pred = model(x)

**Comment by Tps-F** [comment]
PyTorchpartially supports mps.
Unfortunately, not all operations can be done in mps.

**Comment by an-internet-user001** [comment]
So my understanding then is it's not possible to leverage GPUs on Apple Silicone Mac for training.  Is this correct or is it possible?

**Comment by an-internet-user001** [comment]
Was curious if there had been any update to this, or if GPU acceleration on Apple Silicon is still not possible.  Hope all are well!

**Comment by mlatysh** [comment]
Hey guys! Are there any updates on this?

**Comment by Amusesmile** [comment]
How fast is inference when using MPS compared to real time? I've seen CPU inference at about 2x real time.

**Comment by rponeawa** [comment]
Any update?

**Comment by saiisa** [comment]
Training successfully with M4 max, but failed for inference, because failed for feature extraction, because the program cannot recognize Mac GPU ...
